# <center> Güney Marmara ve TRAI
# <center> Python Veri Analizi Eğitimi
# <center> Ödev 2

### BIST30 Endeksinin incelenmesi

1. <i> Yahoo Finance</i> sitesinden 2021 yılına ait BIST30 endeks  değerlerini indiriniz

In [1]:
import yfinance as yf
import matplotlib.pyplot as plt
import plotly.express as px

import warnings
warnings.filterwarnings("ignore")

B30 = yf.download("XU030.IS", start="2021-01-01", end="2021-12-31")
B30

[*********************100%***********************]  1 of 1 completed


Open         High          Low        Close    Adj Close  \
Date                                                                          
2021-01-04  1641.800049  1663.000000  1636.199951  1652.900024  1652.900024   
2021-01-05  1648.599976  1662.099976  1623.900024  1640.400024  1640.400024   
2021-01-06  1653.599976  1659.699951  1645.400024  1657.699951  1657.699951   
2021-01-07  1664.699951  1679.199951  1658.000000  1678.199951  1678.199951   
2021-01-08  1680.900024  1696.599976  1674.900024  1694.400024  1694.400024   
...                 ...          ...          ...          ...          ...   
2021-12-24  2000.300049  2086.800049  1986.800049  2086.800049  2086.800049   
2021-12-27  2093.500000  2143.500000  2063.399902  2089.399902  2089.399902   
2021-12-28  2107.399902  2109.800049  2011.300049  2032.900024  2032.900024   
2021-12-29  2027.500000  2094.600098  2008.199951  2086.300049  2086.300049   
2021-12-30  2109.300049  2119.100098  2037.900024  2040.199951  2040.199951   

             Volume  
Date                 
2021-01-04  2786700  
2021-01-05  4314300  
2021-01-06  3114300  
2021-01-07  3939700  
2021-01-08  3480600  
...             ...  
2021-12-24  1794800  
2021-12-27  2138300  
2021-12-28  1840300  
2021-12-29  1815200  
2021-12-30  1646300  

[238 rows x 6 columns]

2. Veriyi özet metotları ile inceleyiniz (head, info, describe)

In [2]:
B30.head()

Open         High          Low        Close    Adj Close  \
Date                                                                          
2021-01-04  1641.800049  1663.000000  1636.199951  1652.900024  1652.900024   
2021-01-05  1648.599976  1662.099976  1623.900024  1640.400024  1640.400024   
2021-01-06  1653.599976  1659.699951  1645.400024  1657.699951  1657.699951   
2021-01-07  1664.699951  1679.199951  1658.000000  1678.199951  1678.199951   
2021-01-08  1680.900024  1696.599976  1674.900024  1694.400024  1694.400024   

             Volume  
Date                 
2021-01-04  2786700  
2021-01-05  4314300  
2021-01-06  3114300  
2021-01-07  3939700  
2021-01-08  3480600

In [3]:
B30.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 238 entries, 2021-01-04 to 2021-12-30
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       238 non-null    float64
 1   High       238 non-null    float64
 2   Low        238 non-null    float64
 3   Close      238 non-null    float64
 4   Adj Close  238 non-null    float64
 5   Volume     238 non-null    int64  
dtypes: float64(5), int64(1)
memory usage: 13.0 KB


In [4]:
B30.describe()

Open         High          Low        Close    Adj Close  \
count   238.000000   238.000000   238.000000   238.000000   238.000000   
mean   1625.302103  1640.747900  1607.990335  1624.386555  1624.386555   
std     209.287392   219.866592   197.392686   207.108143   207.108143   
min    1381.199951  1415.000000  1301.800049  1397.099976  1397.099976   
25%    1498.800018  1508.199951  1486.274994  1499.000031  1499.000031   
50%    1556.950012  1564.200012  1545.000000  1558.349976  1558.349976   
75%    1648.074982  1659.174957  1632.699951  1642.799988  1642.799988   
max    2585.000000  2677.199951  2397.699951  2528.000000  2528.000000   

             Volume  
count  2.380000e+02  
mean   1.946054e+06  
std    9.024726e+05  
min    1.000000e+02  
25%    1.290600e+06  
50%    1.711100e+06  
75%    2.500850e+06  
max    5.185400e+06

3. Veri temizleme:<br>
	• Tabloda sadece "Adj Close" (düzeltilmiş kapanış) sütununu bırakınız.<br>
	• "Adj Close" sütununu " Kapanış (TRY)" olarak yeniden adlandırınız.

In [5]:
B30 = B30[["Adj Close"]]
B30

Adj Close
Date                   
2021-01-04  1652.900024
2021-01-05  1640.400024
2021-01-06  1657.699951
2021-01-07  1678.199951
2021-01-08  1694.400024
...                 ...
2021-12-24  2086.800049
2021-12-27  2089.399902
2021-12-28  2032.900024
2021-12-29  2086.300049
2021-12-30  2040.199951

[238 rows x 1 columns]

In [6]:
B30.columns = ["Kapanış (TRY)"]
B30

Kapanış (TRY)
Date                     
2021-01-04    1652.900024
2021-01-05    1640.400024
2021-01-06    1657.699951
2021-01-07    1678.199951
2021-01-08    1694.400024
...                   ...
2021-12-24    2086.800049
2021-12-27    2089.399902
2021-12-28    2032.900024
2021-12-29    2086.300049
2021-12-30    2040.199951

[238 rows x 1 columns]

4. Tabloya yeni sütun ekleme:<br>
*   Tabloya yeni bir sütun ekleyerek Kapanış (TRY) değerlerini 1 Ocak 2021 tarihindeki kapanış değerine göre normalize ediniz (yani tüm değerleri 1 Ocak 2021 değerine bölünüz). Bu alana "Normalize" alanı diyebilirsiniz.<br>



In [7]:
B30["Normalize"] = B30["Kapanış (TRY)"]/ B30.iloc[0,0]
B30

Kapanış (TRY)  Normalize
Date                                
2021-01-04    1652.900024   1.000000
2021-01-05    1640.400024   0.992438
2021-01-06    1657.699951   1.002904
2021-01-07    1678.199951   1.015306
2021-01-08    1694.400024   1.025107
...                   ...        ...
2021-12-24    2086.800049   1.262508
2021-12-27    2089.399902   1.264081
2021-12-28    2032.900024   1.229899
2021-12-29    2086.300049   1.262206
2021-12-30    2040.199951   1.234315

[238 rows x 2 columns]

5. Grafik çizdirme ve yorumlama:
*   Tabloyu px.line grafiği olarak görselleştiriniz.<br>
*   BIST30 endeksinin 2021 yılı boyunca değişimini yorumlayınız.

In [8]:
fig_1 = px.line(B30, x=B30.index, y=B30.Normalize,
        width=1000, height=500,
        title = "BIST30 Endeksi 2022 - Normalize (1 ocak 2021)",
        hover_data = {"Normalize": ":.2f", "Kapanış (TRY)": ":.2f",},
        labels = {"Date" : "Tarih", "Normalize" : "BIST 30 Endeksi (TRY)"}
        )
fig_1.show()

# *BITS 30 Endeksi Yıl sonunda 1.23 kat artış olduğu görünmektedir ayrıca yıl içerisinde dengeli bir dağılım varken yıl sonuna doğru ani bir artış gözükmektedir*

6. Değişim alanlarını ekleme ve grafikte koşullu renklendirme:<br>
	• Normalize etiğiniz alan için her satırda bir önceki satırdaki Normalize değeri ile olan farkını hesaplattırınız (bu sütunu "degisim" olarak isimlendirebilirsiniz).
	• Tabloyu px.bar grafiği olarak görselleştiriniz.<br>
	• px.bar grafiğinde endeksin yükseldiği günleri "yeşil", düştüğü günleri "kırmızı" olarak görselleştiriniz.

In [9]:
B30["Degisim"] = B30.Normalize.diff(1)
B30["D_yon"] = B30.Degisim.map(lambda x: "Düşen gün" if x<0 else "Yükselen gün")
B30["Degisim%"] = (B30.Degisim/B30.Normalize.shift(1))

px.bar(B30, y="Degisim%", color="D_yon", color_discrete_map={"Düşen gün": "red", "Yükselen gün": "green"})

7. Histogram çizmek:

*   px.histogram grafiğini çizdiriniz.
*   Bölüm sayısını (nbins) nasıl belirlediniz,açıklayınız.
*   Elde ettiğiniz grafiği yorumlayınız.



nbins değerinin yukarıdaki grafikte gözüktüğü üzere düşen gün ve yükselen günü gruplara ayırırsak düşen gün içerisinde 5 grup ve yükselen grup 3 grup oluştuğunu fark edebiliriz bu yüzden bizim verimizde toplamda 8 grup oluştuğu için nbins 8 seçiyoruz

In [10]:
fig_3 = px.histogram(B30, x="Degisim%", nbins=8)
fig_3.show()